Gridy/Controller/PuzzleViewController.swift

```diff
@@ -35,22 +57,118 @@ class PuzzleViewController: UIViewController, UIGestureRecognizerDelegate {
+        tapGestureRecognizer.delegate = self
+        view.addGestureRecognizer(tapGestureRecognizer)
+        
+        let panGestureRecognizer = UIPanGestureRecognizer(target: self, action: #selector(movePuzzlePieceImageView(_:)))
+        panGestureRecognizer.delegate = self
+        view.addGestureRecognizer(panGestureRecognizer)
+        
+        view.isUserInteractionEnabled = true
+    }
+    
+    @objc func selectPuzzlePieceImageView(_ sender: UILongPressGestureRecognizer) {
+        origin = sender.view?.frame
+    }
+    
+    @objc func movePuzzlePieceImageView(_ sender: UIPanGestureRecognizer) {
+        let location = sender.location(in: self.view)
+        
+        if sender.state == .began {
+            origin = sender.view?.frame
+            //            puzzlePeceState
+        }
+        
+        if sender.state == .changed {
+            sender.view?.frame = CGRect(x: location.x - (sender.view!.frame.width/2),
+                                        y: location.y - (sender.view!.frame.height/2),
+                                        width: sender.view!.frame.width,
+                                        height: sender.view!.frame.height)
+        }
+        
+        if sender.state == .cancelled {
+            sender.view?.frame = origin
+            sender.view?.tag = 0
+        }
+        
+        if sender.state == .ended {
+            for (index, puzzleBlock) in puzzleBlocksViews.enumerated() {
+                if self.view.convert(puzzleBlock.bounds, from: puzzleBlock).contains(location) {
+                    sender.view?.frame = self.view.convert(puzzleBlock.bounds, from: puzzleBlock)
+                    sender.view?.tag = index + 1
+                    break
+                } else {
+                    sender.view?.frame = origin
+                }
+            }
+            for (index, puzzlePiecePlaceholder) in puzzlePiecesPlaceholdersViews.enumerated() {
+                if self.view.convert(puzzlePiecePlaceholder.bounds, from: puzzlePiecePlaceholder).contains(location) {
+                    sender.view?.frame = self.view.convert(puzzlePiecePlaceholder.bounds, from: puzzlePiecePlaceholder)
+                    sender.view?.tag = index + 17
+                    break
+                } else {
+                    sender.view?.frame = origin
+                }
+            }
+        }
+    }
+    
+    func gestureRecognizer(_ gestureRecognizer: UIGestureRecognizer,
+                           shouldRecognizeSimultaneouslyWith otherGestureRecognizer: UIGestureRecognizer)
+        -> Bool {
+            // simultaneous gesture recognition will only be supported for a puzzlePiece
+            for puzzlePiece in puzzlePiecesImageViews {
+                if gestureRecognizer.view != puzzlePiece {
+                    return false
+                }
+            }
+            
+            // neither of the recognized gestures should not be tap gesture
```

Comment:
"neither of the recognized gestures should **not** be tap gesture", the "not" in bold should be removed, shouldn't it?

Reply:
You could also explain why we should return false when it's a tap gesture